### PR TITLE
[Kernel][Defaults] Reuse the file's Hadoop Configuration in the Parquet reader

### PR DESCRIPTION
**Pull request by charlenelyu-db**
#### Which Delta project/connector is this regarding?

- [ ] Spark
- [ ] Standalone
- [ ] Flink
- [x] Kernel
- [ ] Other (fill in here)

## Description

Kernel wraps its own `InputFile` before handing it to `parquet-mr`, so the wrapper is never a `parquet-mr` `HadoopInputFile`. Every entry point that takes an `InputFile` without a configuration therefore falls back to `new HadoopParquetConfiguration()`, which wraps a fresh `Configuration(loadDefaults=true)`. Hadoop loads such a configuration's properties lazily, so the first property read -- which `parquet-mr` issues immediately while building the read options -- scans the classpath for `core-default.xml` and friends under a JVM-global lock. With one of these per Parquet file, concurrent readers serialize on that lock.

Pass the configuration the file is already being read with instead. Its properties are loaded once and memoized on the instance, so the scan disappears. This affects all three call sites that were constructing a fresh configuration: the footer read and the reader build in ParquetFileReader (both per Parquet file), and the footer read in ParquetStatsReader.

## How was this patch tested?

Covered by existing tests

### CODE REVIEW

File: kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/engine/hadoopio/HadoopInputFile.java [hunk 1]
```diff
@@ -47,6 +48,11 @@ public String path() {
     return path.toString();
   }
 
+  /** Returns the Hadoop {@link Configuration} this file is read with. */
+  public Configuration getConfiguration() {
```
Comment (chiinlquah) [inline kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/engine/hadoopio/HadoopInputFile.java]: The surrounding InputFile contract is get-less (path(), length(), newStream()); configuration() reads more consistently with the class's own style.

File: kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ParquetIOUtils.java [hunk 2]
```diff
@@ -30,6 +37,42 @@
...
+   * <p>This matters for concurrency, not just for honoring settings. `parquet-mr` entry points
+   * taking an {@link org.apache.parquet.io.InputFile} that isn't {@code
+   * org.apache.parquet.hadoop.util.HadoopInputFile} fall back to {@code new
+   * HadoopParquetConfiguration()}, which wraps a fresh {@code Configuration(loadDefaults=true)}.
+   * Hadoop loads such a configuration's properties lazily, so the first property read -- which
+   * `parquet-mr` issues immediately while building the read options -- scans the classpath for
+   * `core-default.xml` and friends under a JVM-global lock. This would make concurrent readers
+   * serialize on the lock.
+   *
+   * <p>Reusing the configuration the file is already being read with avoids the scan entirely: its
```
Comment (chiinlquah) [inline kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ParquetIOUtils.java:52]: nit
```
* <p>Reusing the configuration the file is already being read with avoids the scan entirely: its
   * properties are loaded once and memoized on the instance. A non-{@code HadoopInputFile} input
   * carries no Hadoop configuration to reuse, so it falls back to a fresh {@code
   * HadoopParquetConfiguration}
   * 
   ```

File: kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ParquetIOUtils.java [hunk 3]
```diff
@@ -30,6 +37,42 @@
...
+   * properties are loaded once and memoized on the instance.
+   */
+  static ParquetConfiguration parquetConfiguration(InputFile inputFile) {
```
Comment (chiinlquah) [inline kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ParquetIOUtils.java:57]: Let's add a test for it? Seems like every existing parquet test runs with a default/empty conf, so a regression where the wrong conf is threaded (or it is silently ignored) reads identically to correct behavior: reads still succeed
```
scala
// package io.delta.kernel.defaults.internal.parquet (to reach the package-private helper)
test("parquetConfiguration reuses the HadoopInputFile's configuration") {
  val conf = new Configuration(false)
  conf.set("my.sentinel.key", "sentinel-value")
  val input = new HadoopFileIO(conf).newInputFile(somePath, size) // a HadoopInputFile
  val parquetConf = ParquetIOUtils.parquetConfiguration(input)
  assertResult("sentinel-value")(parquetConf.getString("my.sentinel.key", null))
}
```

File: kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ParquetFileReader.java [hunk 4]
```diff
@@ -138,7 +141,7 @@ private void initParquetReaderIfRequired() {
             // no API to do that in the current version of parquet-mr which takes InputFile
             // as input.
             reader =
-                new ParquetReader.Builder<Object>(parquetInputFile) {
+                new ParquetReader.Builder<Object>(parquetInputFile, parquetConf) {
```
Comment (chiinlquah) [inline kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ParquetFileReader.java:147]: Seems like there's a subtle behavior change here worth calling out. With this change, we will now inherit also the non-overridden conf-derived option ie parquet.strings.signed-min-max.enabled (default false) if they were set in the caller, which could prune groups incorrectly if not intended?

**Review by chiinlquah — COMMENTED** [review]
Thanks for the improvements! Some comments

File: kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ParquetFileReader.java (hunk 4 above)
Comment (charlenelyu-db) [reply kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ParquetFileReader.java:147]: Yes it's indeed a behavior change, but I'd argue is the more correct behavior? If a caller has explicitly set a parquet option on the Configuration they pass to HadoopFileIO, we now honor it instead of silently ignoring it and using the default.

I added a comment in the HadoopFileIO.

File: kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ParquetIOUtils.java (hunk 2 above)
Comment (charlenelyu-db) [reply kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ParquetIOUtils.java:52]: Done

File: kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ParquetIOUtils.java (hunk 3 above)
Comment (charlenelyu-db) [reply kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ParquetIOUtils.java:57]: Added test.

File: kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/engine/hadoopio/HadoopFileIO.java [hunk 5]
```diff
@@ -35,6 +35,11 @@
 public class HadoopFileIO implements FileIO {
   private final Configuration hadoopConf;
 
+  /**
+   * @param hadoopConf the Hadoop {@link Configuration} used for all file access through this
+   *     FileIO. Note this configuration is threaded through to the Parquet reader, so any
+   *     parquet-mr options set here rather than being ignored.
```
Comment (chiinlquah) [inline kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/engine/hadoopio/HadoopFileIO.java]: Let's move the comment to the kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ParquetFileReader.java. HadoopFileIO should not care about how all its users use it.

**Review by chiinlquah — APPROVED** [review]
lgtm
